servicetalk-http-netty/src/test/java/io/servicetalk/http/netty/ConnectionContextSocketOptionTest.java

```diff
@@ -0,0 +1,244 @@
+    }
+
+    private <T> void testSocketOption(String name, Class<T> type, Matcher<Object> clientMatcher,
+                                      Matcher<Object> serverMatcher, @Nullable Long idleTimeoutMs) throws Exception {
+        try (ServerContext serverContext = startServer(idleTimeoutMs);
+             BlockingHttpClient client = newClient(serverContext, idleTimeoutMs);
+             BlockingHttpConnection connection = client.reserveConnection(client.get("/"))) {
+
+            SocketOption<T> socketOption = socketOption(name, type);
+            assertThat(connection.connectionContext().socketOption(socketOption), clientMatcher);
+            HttpResponse response = connection.request(connection.get("/")
+                    .setHeader(SOCKET_OPTION_NAME, socketOption.name())
+                    .setHeader(SOCKET_OPTION_TYPE, socketOption.type().getName()));
+            assertThat(response.status(), is(OK));
+            CharSequence value = response.headers().get(SOCKET_OPTION_VALUE);
+            assertThat(value, is(notNullValue()));
+            assertThat(value.toString(), serverMatcher);
+        }
+    }
+
+    private ServerContext startServer(@Nullable Long idleTimeoutMs) throws Exception {
+        final HttpServerBuilder builder = HttpServers.forAddress(localAddress(0))
+                .protocols(protocol.config);
+        if (idleTimeoutMs != null) {
+            builder.socketOption(ServiceTalkSocketOptions.IDLE_TIMEOUT, idleTimeoutMs);
+        }
+        return builder
+                .listenBlockingAndAwait((ctx, request, responseFactory) -> {
+                    try {
+                        String name = request.headers().get(SOCKET_OPTION_NAME).toString();
+                        Class<?> type = Class.forName(request.headers().get(SOCKET_OPTION_TYPE).toString());
+                        SocketOption<?> socketOption = socketOption(name, type);
+                        return responseFactory.ok()
+                                .setHeader(SOCKET_OPTION_VALUE, valueOf(ctx.socketOption(socketOption)));
+                    } catch (Exception e) {
+                        LOGGER.error("Unexpected exception during request processing", e);
+                        return responseFactory.badRequest();
+                    }
+                });
+    }
+
+    private BlockingHttpClient newClient(ServerContext serverContext, @Nullable Long idleTimeoutMs) {
+        SingleAddressHttpClientBuilder<HostAndPort, InetSocketAddress> builder =
+                HttpClients.forSingleAddress(serverHostAndPort(serverContext))
+                        .protocols(protocol.config);
+        if (idleTimeoutMs != null) {
+            builder.socketOption(ServiceTalkSocketOptions.IDLE_TIMEOUT, idleTimeoutMs);
+        }
+        return builder
+                .buildBlocking();
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <T> SocketOption<T> socketOption(String name, Class<T> type) {
```

Comment:
Since you are creating a new `HttpServer` for each test, using the headers indirection is more work (String -> option translations). Instead you can simply pass the socket option and expected value to `startServer` and use that inside the service and then fail the response with the failure message.